### DISCUSSION TITLE
`UserViewSet` vs view per action

### DISCUSSION THREAD

**Discussion by tomwojcik** [sunscrapers/djoser#777] [Polls]
In Djoser 1.x, we utilized a set of views, each handling a specific action. Then came Djoser 2.x, introducing the mighty [UserViewSet](https://github.com/sunscrapers/djoser/blob/641b824353627a73ba901f7e81b86246384d1327/djoser/views.py#L40). IMO it's very simple to register and configure, but for some it might look a bit overwhelming.

For example, [Allauth](https://github.com/pennersr/django-allauth/blob/6790542e3039580ba9e952cc02a8cbef87c09f34/allauth/account/views.py) uses the opposite approach. They have one view per action. It's easier to follow, but results in having much more code.

All feedback welcome!

**Comment by insspb** [comment]
Separate views are much more useful approach. 
If you want to adjust existed viewset and remove some functionalyty you will have a much more code, and it will be completely unclear in some places. 
When we talk about security/auth modules, usually you want to include only parts that you really need, completely excluding any code, that can became accidently 'enabled'.

**Comment by mdefeche** [comment]
I prefer an opt in approach than an opt out. With an opt out approach, newly added endpoints are automatically included which IMHO doesn't make sense from a security point of view

**Comment by GrbavaCigla** [comment]
It's been 5 months since this discussion was started and 3.5 years since the original discussion at #549. I think we won't get more votes than we have got now. I would like to implement the proposed method, would that be merged if maintainers are happy with the solution?

**Reply by haxoza** [reply]
Yes, of course! However we will merge it to separate branch before merging to master. It should be well tested and discussed before releasing it to wider audiance.

Also, if you want to start the implementation please have look to repo history to see how it was structured before the "viewset".

**Reply by GrbavaCigla** [reply]
Excellent, I will start today, you can expect draft pull request soon.

**Reply by GrbavaCigla** [reply]
Just to clarify, we are not doing [this](https://gist.github.com/insspb/e276fb7bb9659a274942153a7f1f3596)?

**Reply by tomwojcik** [reply]
No. We want a similar approach to allauth, so one view per action. This will solve many other requests, such as 

https://github.com/sunscrapers/djoser/issues/416

One other thing we want to avoid is having a mixin for a mixin of a base with a mixin... Applies to both, views and serializers. We will figure it out together during the CR, just wanted to highlight this in advance. 

I've created a branch, feel free to set [`release/3.0.0`](https://github.com/sunscrapers/djoser/compare/master...release/3.0.0) as the target. It's fine if you push WIP in small batches. The only thing that's needed, I'd say, is the CI being 🍏 .

Don't bother with updating the docs for now. Lets have a draft that we will built on top of. 

Thanks for help @GrbavaCigla ! Looking forward for your changes.

**Reply by tomwojcik** [reply]
@GrbavaCigla I will try to address this today. Please push changes if you have them.

**Reply by tomwojcik** [reply]
@GrbavaCigla @haxoza plz CR https://github.com/sunscrapers/djoser/pull/820

**Reply by GrbavaCigla** [reply]
Sorry, I thought this would be less work and I am currently very busy. I skimmed through the changes at #820 and it looks good to me.
